File: CameraCalibration/cameraCalibration.py
#!/usr/bin/env python3
"""Calibrate a camera from checkerboard images with OpenCV 4.14 or 5.0."""

from __future__ import annotations

# argparse gives the teaching script discoverable command-line options.
import argparse
# sys supplies the process argument list and standard error stream.
import sys
# Path anchors bundled assets to this file instead of the caller's directory.
from pathlib import Path
# Sequence describes either a real or test-provided argument list.
from typing import Sequence

# Shared helpers keep both tutorial entry points numerically identical.
from calibration_utils import calibrate_images, print_calibration, validate_calibration


# Resolve the 41 bundled images even when the script starts elsewhere.
PROJECT_DIR = Path(__file__).resolve().parent
# Quote this wildcard at a shell when overriding it so Python, not the shell,
# controls the complete input set.
DEFAULT_IMAGE_PATTERN = str(PROJECT_DIR / "images" / "*.jpg")


def parse_args(arguments: Sequence[str] | None = None) -> argparse.Namespace:
    """Parse command-line options, optionally from a test-provided sequence."""

    # Use the module documentation as the command's short help description.
    parser = argparse.ArgumentParser(description=__doc__)
    # A custom glob lets readers calibrate their own same-sized checkerboards.
    parser.add_argument(
        "--images",
        default=DEFAULT_IMAGE_PATTERN,
        help="Calibration image glob (default: bundled images)",
    )
    # Headless mode avoids all window-system calls on servers and in CI.
    parser.add_argument(
        "--no-display",
        action="store_true",
        help="Skip the interactive checkerboard preview",
    )
    # Validation turns silent numeric assumptions into explicit failures.
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Check version-independent calibration invariants",
    )
    # argparse uses sys.argv when arguments is None.
    return parser.parse_args(arguments)


def run(options: argparse.Namespace) -> None:
    """Execute calibration with already parsed options."""

    # The helper detects corners, refines them, and estimates camera parameters.
    calibration = calibrate_images(
        options.images,
        display=not options.no_display,
    )
    # Print coverage, errors, intrinsics, and distortion for the lesson.
    print_calibration(calibration)
    # Keep validation opt-in for custom datasets with different quality.
    if options.validate:
        validate_calibration(calibration)
        # Tests require an unambiguous marker after every check succeeds.
        print("Validation passed")


def main(arguments: Sequence[str] | None = None) -> int:
    """Run the CLI and convert operational failures into a clean exit status."""

    try:
        # Parse first so argparse can retain its conventional exit code and help.
        options = parse_args(arguments)
        # Execute the real tutorial path rather than a test-only implementation.
        run(options)
    except Exception as error:
        # OpenCV and filesystem errors become a concise command-line diagnostic.
        print(f"Error: {error}", file=sys.stderr)
        return 1
    # Zero tells shells and CTest that all requested work completed.
    return 0


if __name__ == "__main__":
    # SystemExit propagates main's explicit success or failure status.
    raise SystemExit(main())

File: CameraCalibration/cameraCalibration.cpp
// Shared OpenCV 4.14/5.0 calibration logic lives in one teaching-oriented header.
#include "calibration_utils.hpp"

// Filesystem joins the compiled asset directory and wildcard portably.
#include <filesystem>
// The CLI prints help, calibration values, validation markers, and errors.
#include <iostream>
// Invalid options become readable command-line failures.
#include <stdexcept>
// Strings store option names and the selected image pattern.
#include <string>

// Keep path expressions concise in the small option structure.
namespace fs = std::filesystem;

// Hold command-line choices separately from camera-calibration results.
struct Options {
    // CMake embeds the source image directory so any working directory succeeds.
    std::string images =
        (fs::path(TUTORIAL_IMAGE_DIR) / "*.jpg").string();
    // Preserve the original interactive corner preview by default.
    bool display = true;
    // Make regression checks opt-in for normal tutorial exploration.
    bool validate = false;
    // Help exits before loading any image.
    bool show_help = false;
};

// Print the exact interface accepted by this executable.
void print_usage(const char* executable) {
    std::cout
        << "Usage: " << executable
        << " [--images GLOB] [--no-display] [--validate]\n"
        << "\n"
        << "Options:\n"
        << "  --images GLOB  Calibration image pattern (default: bundled images)\n"
        << "  --no-display   Skip interactive checkerboard windows\n"
        << "  --validate     Check calibration invariants\n"
        << "  -h, --help     Show this help text\n";
}

// Convert process arguments into a validated Options value.
Options parse_options(const int argc, char** argv) {
    // Begin with bundled images, display enabled, and validation disabled.
    Options options;
    // argv[0] is the executable name, so option parsing starts at index one.
    for (int index = 1; index < argc; ++index) {
        // Copy the current token so comparisons remain clear and safe.
        const std::string argument = argv[index];
        // The image option must consume exactly one following glob string.
        if (argument == "--images") {
            if (index + 1 >= argc) {
                throw std::invalid_argument(
                    "--images requires a glob pattern"
                );
            }
            // Increment before reading so the value is not parsed as an option.
            options.images = argv[++index];
        } else if (argument == "--no-display") {
            // Headless mode prevents all HighGUI calls in the shared helper.
            options.display = false;
        } else if (argument == "--validate") {
            // Validation checks the result after it has been printed.
            options.validate = true;
        } else if (argument == "-h" || argument == "--help") {
            // Record help instead of exiting inside the parser.
            options.show_help = true;
        } else {
            // Reject typos instead of silently ignoring an intended option.
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    // Return one complete configuration to main.
    return options;
}

// Run the public tutorial entry point.
int main(const int argc, char** argv) {
    try {
        // Validate the command line before accessing files or OpenCV.
        const Options options = parse_options(argc, argv);
        // Help is a successful, side-effect-free request.
        if (options.show_help) {
            print_usage(argv[0]);
            return 0;
        }

        // Detect corners and estimate one camera model from the selected images.
        const tutorial::CalibrationResult calibration =
            tutorial::calibrate_images(options.images, options.display);
        // Print coverage, errors, intrinsics, and distortion coefficients.
        tutorial::print_calibration(calibration);
        // Keep regression checks optional for custom experimental datasets.
        if (options.validate) {
            tutorial::validate_calibration(calibration);
            // CTest requires this marker only after every check succeeds.
            std::cout << "Validation passed\n";
        }
        // Zero signals that all requested work completed successfully.
        return 0;
    } catch (const std::exception& error) {
        // Convert OpenCV, filesystem, option, and validation failures uniformly.
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
